Skip constructor when entity is loaded from database

**Issue type:**

[ ] question
[x] bug report
[x] feature request
[ ] documentation issue

**Database system/driver:**

[x] `cordova`
[x] `mongodb`
[x] `mssql`
[x] `mysql` / `mariadb`
[x] `oracle`
[x] `postgres`
[x] `sqlite`
[x] `sqljs`
[x] `websql`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Create some entity with non-empty constructor. For example:

```ts
@Entity()
export class Post {

    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    private title: string;

    constructor(title: string) {
        if (title === undefined) {
            throw new Error("Invalid title value");
        }

        this.title = title;
    }

}
```

When entity is loaded from db, `Invalid title value` is thrown, because constructor is invoked without args. I think calling constructor must be skipped somehow in this case.

@lodemo I've already tried your solution, but It leads to error: `error TS2564: Property 'title' has no initializer and is not definitely assigned in the constructor.`

--- a/docs/connection-options.md
+++ b/docs/connection-options.md
@@ -72,6 +72,10 @@ Learn more about [Naming strategies](./naming-strategy.md).
 
 * `entityPrefix` - Prefixes with the given string all tables (or collections) on this database connection.
 
+* `entitySkipConstructor` - Indicates if TypeORM should skip constructors when deserializing entities
+  from the database. Note that when you do not call the constructor both private properties and default
+  properties will not operate as expected.
+
 * `dropSchema` - Drops the schema each time connection is being established.
 Be careful with this option and don't use this in production - otherwise you'll lose all production data.
 This option is useful during debug and development.

--- a/src/connection/BaseConnectionOptions.ts
+++ b/src/connection/BaseConnectionOptions.ts
@@ -104,6 +104,11 @@ export interface BaseConnectionOptions {
     readonly entityPrefix?: string;
 
     /**
+     * When creating new Entity instances, skip all constructors when true.
+     */
+    readonly entitySkipConstructor?: boolean;
+
+    /**
      * Extra connection options to be passed to the underlying driver.
      *
      * todo: deprecate this and move all database-specific types into hts own connection options object.

--- a/src/metadata/EmbeddedMetadata.ts
+++ b/src/metadata/EmbeddedMetadata.ts
@@ -87,6 +87,12 @@ export class EmbeddedMetadata {
     embeddeds: EmbeddedMetadata[] = [];
 
     /**
+     * Indicates if the entity should be instantiated using the constructor
+     * or via allocating a new object via `Object.create()`.
+     */
+    isAlwaysUsingConstructor: boolean = true;
+
+    /**
      * Indicates if this embedded is in array mode.
      *
      * This option works only in mongodb.
@@ -189,8 +195,12 @@ export class EmbeddedMetadata {
     /**
      * Creates a new embedded object.
      */
-    create(): any {
-        return new (this.type as any);
+    create(options?: { fromDeserializer?: boolean }): any {
+        if (!options?.fromDeserializer || this.isAlwaysUsingConstructor) {
+            return new (this.type as any);
+        } else {
+            return Object.create(this.type.prototype);
+        }
     }
 
     // ---------------------------------------------------------------------
@@ -211,6 +221,11 @@ export class EmbeddedMetadata {
         this.uniquesFromTree = this.buildUniquesFromTree();
         this.relationIdsFromTree = this.buildRelationIdsFromTree();
         this.relationCountsFromTree = this.buildRelationCountsFromTree();
+
+        if (connection.options.entitySkipConstructor) {
+            this.isAlwaysUsingConstructor = !connection.options.entitySkipConstructor;
+        }
+
         return this;
     }
 

--- a/src/metadata/EntityMetadata.ts
+++ b/src/metadata/EntityMetadata.ts
@@ -188,6 +188,12 @@ export class EntityMetadata {
     isJunction: boolean = false;
 
     /**
+     * Indicates if the entity should be instantiated using the constructor
+     * or via allocating a new object via `Object.create()`.
+     */
+    isAlwaysUsingConstructor: boolean = true;
+
+    /**
      * Indicates if this entity is a tree, what type of tree it is.
      */
     treeType?: TreeType;
@@ -525,11 +531,16 @@ export class EntityMetadata {
     /**
      * Creates a new entity.
      */
-    create(queryRunner?: QueryRunner): any {
+    create(queryRunner?: QueryRunner, options?: { fromDeserializer?: boolean }): any {
         // if target is set to a function (e.g. class) that can be created then create it
         let ret: any;
         if (this.target instanceof Function) {
-            ret = new (<any> this.target)();
+            if (!options?.fromDeserializer || this.isAlwaysUsingConstructor) {
+                ret = new (<any> this.target)();
+            } else {
+                ret = Object.create(this.target.prototype);
+            }
+
             this.lazyRelations.forEach(relation => this.connection.relationLoader.enableLazyLoad(relation, ret, queryRunner));
             return ret;
         }
@@ -781,6 +792,8 @@ export class EntityMetadata {
     build() {
         const namingStrategy = this.connection.namingStrategy;
         const entityPrefix = this.connection.options.entityPrefix;
+        const entitySkipConstructor = this.connection.options.entitySkipConstructor;
+
         this.engine = this.tableMetadataArgs.engine;
         this.database = this.tableMetadataArgs.type === "entity-child" && this.parentEntityMetadata ? this.parentEntityMetadata.database : this.tableMetadataArgs.database;
         if (this.tableMetadataArgs.schema) {
@@ -814,6 +827,10 @@ export class EntityMetadata {
         this.tablePath = this.buildTablePath();
         this.schemaPath = this.buildSchemaPath();
         this.orderBy = (this.tableMetadataArgs.orderBy instanceof Function) ? this.tableMetadataArgs.orderBy(this.propertiesMap) : this.tableMetadataArgs.orderBy; // todo: is propertiesMap available here? Looks like its not
+
+        if (entitySkipConstructor !== undefined) {
+            this.isAlwaysUsingConstructor = !entitySkipConstructor;
+        }
 
         this.isJunction = this.tableMetadataArgs.type === "closure-junction" || this.tableMetadataArgs.type === "junction";
         this.isClosureJunction = this.tableMetadataArgs.type === "closure-junction";

--- a/src/query-builder/transformer/DocumentToEntityTransformer.ts
+++ b/src/query-builder/transformer/DocumentToEntityTransformer.ts
@@ -28,7 +28,7 @@ export class DocumentToEntityTransformer {
     }
 
     transform(document: any, metadata: EntityMetadata) {
-        const entity: any = metadata.create();
+        const entity: any = metadata.create(undefined, { fromDeserializer: true });
         let hasData = false;
 
         // handle _id property the special way
@@ -87,7 +87,7 @@ export class DocumentToEntityTransformer {
 
                 if (embedded.isArray) {
                     entity[embedded.propertyName] = (document[embedded.prefix] as any[]).map((subValue: any, index: number) => {
-                        const newItem = embedded.create();
+                        const newItem = embedded.create({ fromDeserializer: true });
                         embedded.columns.forEach(column => {
                             newItem[column.propertyName] = subValue[column.databaseNameWithoutPrefixes];
                         });
@@ -96,15 +96,15 @@ export class DocumentToEntityTransformer {
                     });
 
                 } else {
-                    if (embedded.embeddeds.length && !entity[embedded.propertyName]) 
-                        entity[embedded.propertyName] = embedded.create();
-                    
+                    if (embedded.embeddeds.length && !entity[embedded.propertyName])
+                        entity[embedded.propertyName] = embedded.create({ fromDeserializer: true });
+
                     embedded.columns.forEach(column => {
                         const value = document[embedded.prefix][column.databaseNameWithoutPrefixes];
                         if (value === undefined) return;
 
                         if (!entity[embedded.propertyName])
-                            entity[embedded.propertyName] = embedded.create();
+                            entity[embedded.propertyName] = embedded.create({ fromDeserializer: true });
 
                         entity[embedded.propertyName][column.propertyName] = value;
                     });

--- a/src/query-builder/transformer/PlainObjectToNewEntityTransformer.ts
+++ b/src/query-builder/transformer/PlainObjectToNewEntityTransformer.ts
@@ -66,7 +66,7 @@ export class PlainObjectToNewEntityTransformer {
 
                         // if such item already exist then merge new data into it, if its not we create a new entity and merge it into the array
                         if (!objectRelatedValueEntity) {
-                            objectRelatedValueEntity = relation.inverseEntityMetadata.create();
+                            objectRelatedValueEntity = relation.inverseEntityMetadata.create(undefined, { fromDeserializer: true });
                             entityRelatedValue.push(objectRelatedValueEntity);
                         }
 
@@ -86,7 +86,7 @@ export class PlainObjectToNewEntityTransformer {
                     }
 
                     if (!entityRelatedValue) {
-                        entityRelatedValue = relation.inverseEntityMetadata.create();
+                        entityRelatedValue = relation.inverseEntityMetadata.create(undefined, { fromDeserializer: true });
                         relation.setEntityValue(entity, entityRelatedValue);
                     }
 

--- a/src/query-builder/transformer/RawSqlResultsToEntityTransformer.ts
+++ b/src/query-builder/transformer/RawSqlResultsToEntityTransformer.ts
@@ -102,7 +102,7 @@ export class RawSqlResultsToEntityTransformer {
             if (discriminatorMetadata)
                 metadata = discriminatorMetadata;
         }
-        let entity: any = this.expressionMap.options.indexOf("create-pojo") !== -1 ? {} : metadata.create(this.queryRunner);
+        let entity: any = this.expressionMap.options.indexOf("create-pojo") !== -1 ? {} : metadata.create(this.queryRunner, { fromDeserializer: true });
 
         // get value from columns selections and put them into newly created entity
         const hasColumns = this.transformColumns(rawResults, alias, entity, metadata);

--- a/test/functional/entity-metadata/entity-metadata-create.ts
+++ b/test/functional/entity-metadata/entity-metadata-create.ts
@@ -1,0 +1,81 @@
+import "reflect-metadata";
+import { Connection } from "../../../src/connection/Connection";
+import { expect } from "chai";
+import { closeTestingConnections, createTestingConnections, reloadTestingDatabases } from "../../utils/test-utils";
+import { TestCreate } from "./entity/TestCreate";
+
+describe("entity-metadata > create", () => {
+    describe("without entitySkipConstructor", () => {
+        let connections: Connection[];
+        before(async () => connections = await createTestingConnections({
+                enabledDrivers: [ "sqlite" ],
+                entities: [
+                    TestCreate
+                ]
+            })
+        );
+
+        beforeEach(() => reloadTestingDatabases(connections));
+        after(() => closeTestingConnections(connections));
+
+        it("should call the constructor when creating an object", () => Promise.all(connections.map(async connection => {
+            const entity = connection.manager.create(TestCreate);
+
+            expect(entity.hasCalledConstructor).to.be.true;
+        })))
+
+        it("should set the default property values", () => Promise.all(connections.map(async connection => {
+            const entity = connection.manager.create(TestCreate);
+
+            expect(entity.foo).to.be.equal("bar");
+        })))
+
+        it("should call the constructor when retrieving an object", () => Promise.all(connections.map(async connection => {
+            const repo = connection.manager.getRepository(TestCreate);
+
+            const { id } = await repo.save({ foo: "baz" });
+
+            const entity = await repo.findOneOrFail(id);
+
+            expect(entity.hasCalledConstructor).to.be.true;
+        })))
+    })
+
+    describe("with entitySkipConstructor", () => {
+        let connections: Connection[];
+        before(async () => connections = await createTestingConnections({
+            enabledDrivers: [ "sqlite" ],
+            entities: [
+                TestCreate
+            ],
+            driverSpecific: {
+                entitySkipConstructor: true,
+            }
+        }));
+
+        beforeEach(() => reloadTestingDatabases(connections));
+        after(() => closeTestingConnections(connections));
+
+        it("should call the constructor when creating an object", () => Promise.all(connections.map(async connection => {
+            const entity = connection.manager.create(TestCreate);
+
+            expect(entity.hasCalledConstructor).to.be.true;
+        })))
+
+        it("should set the default property values when creating an object", () => Promise.all(connections.map(async connection => {
+            const entity = connection.manager.create(TestCreate);
+
+            expect(entity.foo).to.be.equal("bar");
+        })))
+
+        it("should not call the constructor when retrieving an object", () => Promise.all(connections.map(async connection => {
+            const repo = connection.manager.getRepository(TestCreate);
+
+            const { id } = await repo.save({ foo: "baz" });
+
+            const entity = await repo.findOneOrFail(id);
+
+            expect(entity.hasCalledConstructor).not.to.be.true;
+        })))
+    })
+})

--- a/test/functional/entity-metadata/entity/TestCreate.ts
+++ b/test/functional/entity-metadata/entity/TestCreate.ts
@@ -1,0 +1,18 @@
+import { Column } from "../../../../src/decorator/columns/Column";
+import { Entity } from "../../../../src/decorator/entity/Entity";
+import { PrimaryGeneratedColumn } from "../../../../src";
+
+@Entity()
+export class TestCreate {
+    constructor() {
+        this.hasCalledConstructor = true;
+    }
+
+    hasCalledConstructor = false;
+
+    @PrimaryGeneratedColumn()
+    id: number;
+
+    @Column()
+    foo: string = 'bar';
+}

